src/cli/virtualfs.ts

```diff
@@ -51,157 +43,106 @@ export function vfsUri(
-    fs.stat(pathInWorkSpace, (err, stats) => {
-      if (err) {
-        throw err;
-      }
-      fileType = FileType.File;
-      createTime = stats.ctimeMs;
-      modifiedTime = stats.mtimeMs;
-      fileSize = stats.size;
-    });
-
-    return {
-      type: fileType,
-      ctime: createTime,
-      mtime: modifiedTime,
-      size: fileSize,
-    };
+  stat(uri: Uri): FileStat | Thenable<FileStat> {
+    return this.ensureStat(uri);
   }
 
-  readDirectory(_uri: Uri): [string, FileType][] | Thenable<[string, FileType][]> {
-    return this.readDirectoryAsync();
-  }
+  private ensureStat(uri: Uri): VFSFileStat {
+    if (!this.fileStats.has(uri.toString())) {
+      this.fileStats.set(uri.toString(), new VFSFileStat());
+    }
 
-  /**
-   * Add the directory if missing.
-   *
-   * Return an array of the array of `[path, fileType]` for the files in the dir.
-   */
-  async readDirectoryAsync(): Promise<[string, FileType][]> {
-    const files: [string, FileType][] = [];
-    await this.createDirectoryAsync(null);
-    const dir = await getFilePathAsync(this.yamlDirName, null);
-    fs.readdirSync(dir).forEach((localFile) => {
-      files.push([path.join(dir, localFile), FileType.File]);
-    });
-    return files;
+    const stat = this.fileStats.get(uri.toString());
+    stat.changeStat(stat.size + 1);
+
+    return stat;
   }
 
-  createDirectory(_uri: Uri): void | Thenable<void> {
-    return this.createDirectoryAsync(_uri);
+  readFile(uri: Uri): Uint8Array | Thenable<Uint8Array> {
+    return this.readFileAsync(uri);
   }
 
-  /**
-   * If directory does not already exist, create it.
-   * @param _uri Ignored. May be used in the future.
-   */
-  async createDirectoryAsync(_uri: Uri): Promise<void> {
-    const dir = await getFilePathAsync(this.yamlDirName, null);
+  async writeFile(uri: Uri, content: Uint8Array): Promise<void> {
+    const tempPath = os.tmpdir();
+    const fsPath = path.join(tempPath, uri.fsPath);
+    await fsx.ensureFile(fsPath);
+    await fsx.writeFile(fsPath, content);
+    await this.updateK8sResource(fsPath);
+    await this.unlinkFsPath(fsPath);
+    servingDataProvider.refresh();
+  }
 
-    // eslint-disable-next-line no-useless-catch
+  // eslint-disable-next-line class-methods-use-this
+  async updateK8sResource(fsPath: string): Promise<void> {
     try {
-      if (!fs.existsSync(dir)) {
-        fs.mkdirSync(dir);
+      // push the updated YAML back to the cluster
+      const result: CliExitData = await this.knExecutor.execute(KnAPI.applyYAML(fsPath, { override: false }));
+      // Delete the yaml that was pushed if there was no error
+      if (result.error) {
+        // eslint-disable-next-line no-console, @typescript-eslint/restrict-template-expressions
+        // eslint-disable-next-line @typescript-eslint/no-floating-promises
+        window.showErrorMessage(`updateServiceFromYaml result.error = ${getStderrString(result.error)}`);
+        await this.unlinkFsPath(fsPath);
```

Comment:
these two lines can be removed. You are throwing an error and it will get inside the catch which already handles the error message and unlink the file. If you want to keep it the `showErrorMessage` should have the `await`? :thinking: